control/netns_utils.go

```diff
@@ -0,0 +1,184 @@
+package control
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"runtime"
+	"sync"
+
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	indieNetns netns.NsHandle
+	once       sync.Once
+)
+
+func WithIndieNetns(f func() error) (err error) {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	hostNetns, err := netns.Get()
+	if err != nil {
+		return fmt.Errorf("Failed to get host netns: %v", err)
+	}
+	defer netns.Set(hostNetns)
+
+	ns, err := GetIndieNetns()
+	if err != nil {
+		return
+	}
+	if err = netns.Set(ns); err != nil {
+		return fmt.Errorf("Failed to switch to daens: %v", err)
+	}
+
+	return f()
+}
+
+func GetIndieNetns() (_ netns.NsHandle, err error) {
+	if indieNetns != 0 {
+		return indieNetns, nil
+	}
+
+	once.Do(func() {
+		err = setupIndieNetns()
```

Comment:
if err occurs, indieNetns will never be set and GetIndieNetns will return 0, nil that is not expected.

Reply:
Latest version resets once on error:

```go
	once.Do(func() {
		daeNetns, err = setupDaeNetns()
		if err != nil {
			once = sync.Once{}
		}
	})
```